### PR TITLE
Don't flush dirty pages when closing files mapped with DELETE_ON_CLOSE

### PR DESCRIPTION
**Pull request by chrisvest**

### CODE REVIEW

**Comment by chrisvest** [comment]
Please take a look, @tinwelint. You too, @ragadeeshu if you like.

File: community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java [hunk 1]
```diff
@@ -261,13 +261,68 @@ public void flushAndForce( IOLimiter limiter ) throws IOException
...
+
+                        if ( !page.tryExclusiveLock() )
+                        {
+                            continue;
+                        }
+                        if ( page.isBoundTo( swapper, filePageId ) && page.isDirty() )
+                        {
+                            // The page is still bound to the expected file and file page id after we locked it,
+                            // so we didn't race with eviction and faulting, and the page is dirty.
+                            // So we add it to our IO vector.
```
Comment (burqen) [inline community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java]: Nice comment but I don't think the end of it describes what we do.

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java [hunk 2]
```diff
@@ -3866,13 +3891,95 @@ public void fileMappedWithDeleteOnCloseMustNotExistAfterLastUnmap() throws Excep
...
+        try ( PagedFile pf = cache.map( file, filePageSize, DELETE_ON_CLOSE );
+              WritableByteChannel channel = pf.openWritableByteChannel() )
+        {
+            generateFileWithRecords( channel, recordCount, recordSize );
+        }
+        assertThat( flushCounter.get(), lessThan( recordCount / recordsPerFilePage ) );
+    }
+
+    @Test
+    public void mustFlushAllDirtyPagesWhenClosingPagedFileThatIsNotMappedWithDeleteOnClose() throws Exception
```
Comment (burqen) [inline community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java:3918]: Is this test not testing the same thing as `dirtyPagesMustBeFlushedWhenThePagedFileIsClosed` (also added in this PR)

**Review by burqen — COMMENTED** [review]
Some curiosity driven comments

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java (hunk 2 above)
Comment (chrisvest) [reply community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java:3918]: They are related but make different observations, and are symmetric with different tests.

**Comment by chrisvest** [comment]
@burqen fixed your comment on the comment :P

**Comment by chrisvest** [comment]
@burqen Finally green 💯

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java [hunk 3]
```diff
@@ -5315,17 +5386,16 @@ public void streamFilesRecursiveMustNotSeeFilesRenamedIntoBaseDirectory() throws
...
-            File file = handle.getFile();
+            File file = fh.getFile();
             observedFiles.add( file );
             if ( file.equals( x ) )
             {
-                handle.rename( target );
+                handleRename( target ).accept( fh );
             }
-        }
+        });
```
Comment (burqen) [inline community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java]: Code style

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java [hunk 4]
```diff
@@ -5336,17 +5406,16 @@ public void streamFilesRecursiveMustNotSeeFilesRenamedIntoSubDirectory() throws
...
-            File file = handle.getFile();
+            File file = fh.getFile();
             observedFiles.add( file );
             if ( file.equals( a ) )
             {
-                handle.rename( target );
+                handleRename( target ).accept( fh );
             }
-        }
+        });
```
Comment (burqen) [inline community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java]: Code style

**Comment by chrisvest** [comment]
@burqen This now fails the new check style rules. I will let you know when it's updated.

**Comment by chrisvest** [comment]
@burqen Updated.

**Review by burqen — COMMENTED** [review]
Two additional code style comment. Otherwise ready for merge.

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java (hunk 4 above)
Comment (chrisvest) [reply community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java]: Except checkstyle is not complaining about it?

**Comment by burqen** [comment]
Our checkstyle checker does not check everything that is in our checkstyle
rule. But intellij disagrees with formatting.

File: community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java (hunk 4 above)
Comment (burqen) [reply community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java]: Our checkstyle checker does not check everything that is in our checkstyle
rule. But intellij disagrees with formatting.

**Comment by chrisvest** [comment]
@burqen Fine, updated.

**Review by burqen — APPROVED** [review]

**Pull request merged**